### DISCUSSION TITLE
Metadata Management Replacement

### DISCUSSION THREAD

**Discussion by nichwall** [advplyr/audiobookshelf#2479] [General]
As briefly discussed in Discord, here is an overview of my idea for changing the metadata tools page. Please comment on anything that is confusing or doesn't make sense. I was going to make a video but then decided it may make discussion harder as things change.

## Overview
My idea for the new metadata/encoder manager is to move it from being a dedicated page to being a tab in the “Edit” modal (replacing the Tools tab). This tab can exist for the following items:
- Library
- Book
- Podcast

![PXL_20240103_052914051.jpg](https://github.com/advplyr/audiobookshelf/assets/5686638/7a730cfe-8e6f-410e-b88d-a250a43fc900)

The tab includes the main sections: text information, encode options, and metadata information.

### Information Section
The text information in the upper left is brief information about encoding and embedding metadata that is currently located in the bullet points on the existing Manager page. This section can link to a guide because explaining everything concisely may be difficult. But, if anything is not intuitive, translations of the web documentation do not exist.

I'm not sure whether the “options” or “metadata” should be next, but I have sketched it as “options”.

### Encode options
In the upper right, there is a toggle to switch between “encode/convert” and “embed”. If the “encode/convert” is selected, then the options boxes appear as shown in the sketch (but can be rearranged to fit better). There is a checkbox to “use library defaults” which is enabled by default. If this box is checked, all of the drop downs are greyed out and set to the default values based on the library settings. If the checkmark is enabled again after changing settings, the values reset to the defaults.

The encoder settings which can be configured are as follows, with the default value for a newly created audiobook library marked.
- Bitrate (enum)
  - Max (default) = bitrate of output is set to the highest bitrate of the input files
  - Min = bitrate of output is set to the lowest bitrate of the input files
  - Fixed = bitrate of output is the value entered in the box. If set to max or min, the box displays what bitrate will be used and is greyed out/does not accept user input.
- Channels (enum)
  - Keep (default) = If any input files have stereo, use stereo. Otherwise use mono
  - Mono = use a single output track
  - Stereo = use two output tracks
- Codec (enum)
  - AAC (default)
  - Opus
  - Others?
- Extension (enum)
  - M4b (default)
  - M4a
  - Mp3
  - Others?
- Combine (toggle)
  - True (default) = combine input files into a single output file
  - False = keep each file separate in output.

Some logic would need to be done to ensure the codec and container are compatible.

If the toggle at the top is set to “embed”, then the encoding settings are hidden.

### Metadata Table
The expandable metadata table would include a list of what fields would be mapped from ABS (by container), but I don't think their value needs to be shown here since all of that information is included in other modal tabs.

## Modifications:
Earlier I mentioned this layout can be used for libraries, books, and podcasts. This section describes the differences between each modal.

- **Book**: as described above. Not persistent.

- **Audiobook library**: hide “embed” toggle, since it will only be encode options. Remove “use library defaults” checkbox. Persistent settings.

- **Podcast library**: replace “combine” toggle with “autoconvert” which will automatically perform the encode after downloading an episode (like to opus or a lower bitrate). Remove “user library defaults” checkbox. Persistent settings. Does not apply to already downloaded files.

- **Podcast**: same as podcast library, by overrides the library settng. Persistent settings. Does not apply to already downloaded files, but could have an indicator with an "encode" button in the file list in the Files tab if the file doesn't match the podcast/library settings.

## Main concerns:
I think it is better for the tab to look similar so users can expect similar behavior in each of the models, but I am concerned that users will mix them up if they all look the same.

**Comment by nichwall** [comment]
Additional thoughts:

A separate page is no longer needed because users can navigate away from the page while an encode continues working in the background. An additional "encode queue" page (can be based on the podcast download queue page) should be added whenever that functionality is added, but doesn't need to be created yet.

**Reply by nichwall** [reply]
Eventually, once this is all implemented, it would be cool if there was a filter for "items in library which don't match the default library (or podcast) setting" similar to the existing multi/single track book filter to more easily find books which have been missed. I don't think there should be a "encode everything on server" bulk button yet. Maybe eventually, once it's been determined that everything is working.

**Comment by mdeeter** [comment]
I love the ideas presented.

Perhaps an option (or a configuration setting) to save original file(s) in a backup folder/location?

**Reply by nichwall** [reply]
Ah, thanks for taking the time to review, and mentioning the option to backup the files. :)
Yes, a toggle for that should be included for that (like there currently is).

It may be handy to have an "automatic cleaning" function at some point, but I think keeping it as the user manually cleans the backup cache for now.

**Comment by advplyr** [comment]
I think this looks great! 
Underneath the encoding options we could show what the current values are similar to how we do it in the match tab.

**Reply by nichwall** [reply]
Thanks for taking the time to review!

I've slowly been working on getting the initial library settings tab but out, but I'm really slow with JS/Vue. Right now I'm only focusing on getting the Audiobook library tab built out and updating the existing encode page to use the min/max bitrates.

I don't mind if my current PR gets ignored because there's a cleaner way to do it, but I'll keep working on it in the meantime (I will eventually need help getting the UI all nice and consistent since I've just been doing it by trial and error).

https://github.com/advplyr/audiobookshelf/pull/2472
